### PR TITLE
Use proper executor for failing requests when connection closes

### LINKED ISSUE #109225: SearchableSnapshotsIntegTests.testCreateAndRestoreSearchableSnapshot failing
### CI Link

https://gradle-enterprise.elastic.co/s/ivesuuq5fvjji

### Repro line

./gradlew ':x-pack:plugin:searchable-snapshots:internalClusterTest' --tests "org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotsIntegTests.testCreateAndRestoreSearchableSnapshot" -Dtests.seed=613FCFFF93F38F06 -Dtests.locale=es-MX -Dtests.timezone=Africa/Lagos -Druntime.java=22

### Does it reproduce?

Not reliably

### Applicable branches

lucene_snapshot_9_11

### Failure history

_No response_

### Failure excerpt

```
java.lang.AssertionError: cannot complete future on thread Thread[#199,elasticsearch[node_s2][generic][T#3],5,TGRP-SearchableSnapshotsIntegTests] with waiter on thread Thread[#316,elasticsearch[node_s2][generic][T#11],5,TGRP-SearchableSnapshotsIntegTests], could deadlock if pool was full
--
  | at java.base/java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:269)
  | at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:756)
  | at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(AbstractQueuedSynchronizer.java:1126)
  | at org.elasticsearch.action.support.PlainActionFuture$Sync.get(PlainActionFuture.java:262)
  | at org.elasticsearch.action.support.PlainActionFuture.get(PlainActionFuture.java:74)
  | at org.elasticsearch.common.util.concurrent.FutureUtils.get(FutureUtils.java:65)
  | at org.elasticsearch.action.support.PlainActionFuture.actionGet(PlainActionFuture.java:171)
  | at org.elasticsearch.xpack.searchablesnapshots.cache.blob.BlobStoreCacheService.get(BlobStoreCacheService.java:126)
  | at org.elasticsearch.xpack.searchablesnapshots.store.SearchableSnapshotDirectory.getCachedBlob(SearchableSnapshotDirectory.java:673)
  | at org.elasticsearch.xpack.searchablesnapshots.store.input.MetadataCachingIndexInput.readWithBlobCache(MetadataCachingIndexInput.java:280)
  | at org.elasticsearch.xpack.searchablesnapshots.store.input.MetadataCachingIndexInput.readInternal(MetadataCachingIndexInput.java:615)
  | at org.elasticsearch.blobcache.common.BlobCacheBufferedIndexInput.refill(BlobCacheBufferedIndexInput.java:304)
  | at org.elasticsearch.blobcache.common.BlobCacheBufferedIndexInput.readByte(BlobCacheBufferedIndexInput.java:57)
  | at org.apache.lucene.store.BufferedChecksumIndexInput.readByte(BufferedChecksumIndexInput.java:39)
  | at org.apache.lucene.codecs.CodecUtil.readBEInt(CodecUtil.java:668)
  | at org.apache.lucene.codecs.CodecUtil.checkHeader(CodecUtil.java:185)
  | at org.apache.lucene.codecs.CodecUtil.checkIndexHeader(CodecUtil.java:254)
  | at org.apache.lucene.codecs.lucene94.Lucene94FieldInfosFormat.read(Lucene94FieldInfosFormat.java:138)
  | at org.apache.lucene.index.IndexWriter.readFieldInfos(IndexWriter.java:1259)
  | at org.apache.lucene.index.IndexWriter.getFieldNumberMap(IndexWriter.java:1278)
  | at org.apache.lucene.index.IndexWriter.<init>(IndexWriter.java:1125)
  | at org.elasticsearch.index.store.Store.newTemporaryAppendingIndexWriter(Store.java:1546)
  | at org.elasticsearch.index.store.Store.associateIndexWithNewTranslog(Store.java:1452)
  | at org.elasticsearch.index.shard.StoreRecovery.bootstrap(StoreRecovery.java:620)
  | at org.elasticsearch.index.shard.StoreRecovery.lambda$restore$17(StoreRecovery.java:586)
  | at org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:245)
  | at org.elasticsearch.action.support.SubscribableListener$SuccessResult.complete(SubscribableListener.java:382)
  | at org.elasticsearch.action.support.SubscribableListener.tryComplete(SubscribableListener.java:302)
  | at org.elasticsearch.action.support.SubscribableListener.addListener(SubscribableListener.java:205)
  | at org.elasticsearch.action.support.SubscribableListener.lambda$andThen$0(SubscribableListener.java:469)
  | at org.elasticsearch.action.ActionListener.run(ActionListener.java:433)
  | at org.elasticsearch.action.support.SubscribableListener.newForked(SubscribableListener.java:128)
  | at org.elasticsearch.action.support.SubscribableListener.andThen(SubscribableListener.java:469)
  | at org.elasticsearch.action.support.SubscribableListener.andThen(SubscribableListener.java:433)
  | at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:583)
  | at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:312)
  | at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:2481)
  | at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$32(IndexShard.java:3268)
  | at org.elasticsearch.action.ActionRunnable$4.doRun(ActionRunnable.java:100)
  | at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:984)
  | at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
  | at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
  | at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
  | at java.base/java.lang.Thread.run(Thread.java:1570)
```

### PR DESCRIPTION
**Pull request by henningandersen**
Now use the executor declared by the handler rather than generic, since using generic is trappy wrt. deadlocks.

Closes #109225

### CODE REVIEW

**Comment by henningandersen** [comment]
Opening as draft to see what CI and devs thinks before adding more tests.

**Comment by henningandersen** [comment]
ci/part-2 failure is #109237

**Comment by henningandersen** [comment]
This is now ready for reviews.

**Review by DaveCTurner — COMMENTED** [review]
Overall approach looks good, I left a handful of smaller comments.

File: server/src/main/java/org/elasticsearch/transport/TransportService.java [hunk 1]
```diff
@@ -1342,36 +1357,52 @@ public void onConnectionClosed(Transport.Connection connection) {
...
-            }
+            TransportResponseHandler<?> handler = holderToNotify.handler();
+            // we used to fork to a different thread always to avoid stack overflows, but we avoid doing that now, expecting handlers
+            // to handle that themselves instead.
+            final var executor = handler.executor();
+            if (executor == EsExecutors.DIRECT_EXECUTOR_SERVICE) {
+                if (enableStackOverflowAvoidance == false) {
+                    handler.handleException(exception);
+                } else {
+                    threadPool.generic().submit(new AbstractRunnable() {
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/transport/TransportService.java:1379]: It is slightly annoying in that it asserts that the handler is not using direct executor, hence i did this route. I can make it work ofc, let me give it a shot.

[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/transport/TransportService.java:1379]: Yeah sure I meant not for the direct case, just if `enableStackOverflowAvoidance`

[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/transport/TransportService.java]: Yeah, absolutely, the handler verification is still annoying, but made it slightly more optional when fixing this here:

f93cc1c4c73488930791eaf6c164c29c2983d087

File: server/src/test/java/org/elasticsearch/transport/TransportServiceLifecycleTests.java [hunk 2]
```diff
@@ -217,6 +219,56 @@ public void testInternalSendExceptionCompletesHandlerOnCallingThreadIfTransportS
...
+                new ActionListenerResponseHandler<>(ActionListener.assertOnce(future.delegateResponse((l, e) -> {
+                    assertThat(EsExecutors.executorName(Thread.currentThread()), equalTo(expectedExecutor));
+                    if (expectedExecutor == null) {
+                        assertSame(testThread, Thread.currentThread());
+                    }
+                    l.onFailure(e);
+                })), unusedReader(), executor)
+            );
+            nodeA.transportService.onConnectionClosed(connection);
+            expectThrows(ExecutionException.class, NodeDisconnectedException.class, () -> future.get(10, TimeUnit.SECONDS));
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/transport/TransportServiceLifecycleTests.java]: Yeah, that is better, see: f34ef8f8ab9d984a56d0c1d201876cfe52369c01

File: server/src/test/java/org/elasticsearch/transport/TransportServiceLifecycleTests.java [hunk 3]
```diff
@@ -217,6 +219,56 @@ public void testInternalSendExceptionCompletesHandlerOnCallingThreadIfTransportS
...
+            "[transport.enable_stack_protection] setting was deprecated in Elasticsearch and will be removed in a future release."
+        );
+    }
+
+    private void onConnectionClosedUsesHandlerExecutor(Settings settings, String executorName, String expectedExecutor) {
+        try (var nodeA = new TestNode("node-A", settings)) {
+            final var testThread = Thread.currentThread();
+            final var future = new PlainActionFuture<TransportResponse.Empty>();
+            Executor executor = nodeA.getExecutor(executorName);
+            Transport.Connection connection = nodeA.getDevNullConnection();
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/transport/TransportServiceLifecycleTests.java:255]: Also, the `sendRequest` looks up the discovery node, hence I left it on the `TestNode` to be able to use the `transportService.localNode()`, just like `getThrowingConnection`. I could do it outside, but I am not sure I see a good reason why?

File: server/src/main/java/org/elasticsearch/transport/TransportService.java [hunk 4]
```diff
@@ -90,6 +91,17 @@ public class TransportService extends AbstractLifecycleComponent
     public static final String DIRECT_RESPONSE_PROFILE = ".direct";
     public static final String HANDSHAKE_ACTION_NAME = "internal:transport/handshake";
 
+    /**
+     * Undocumented on purpose, may be removed at any time. Only use this if instructed to do so, can have other unintended consequences
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/transport/TransportService.java:95]: Thanks, registered it in 1027967d81c5f4e692a1a21d41621b48579a32b9

I was slightly conflicted on what to do with testing, but came up with this: c4c3164028a7847f34400c8fe0bac846fdabdfa9

File: server/src/main/java/org/elasticsearch/transport/TransportService.java [hunk 5]
```diff
@@ -1342,36 +1357,35 @@ public void onConnectionClosed(Transport.Connection connection) {
...
 
-            @Override
-            public void onFailure(Exception e) {
-                assert false : e;
-                logger.warn(() -> "failed to notify response handler on connection close [" + connection + "]", e);
+            TransportResponseHandler<?> handler = holderToNotify.handler();
+            // we used to fork to a different thread always to avoid stack overflows, but we avoid doing that now, expecting handlers
+            // to handle that themselves instead.
+            var executor = handler.executor();
+            if (executor == EsExecutors.DIRECT_EXECUTOR_SERVICE) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/TransportService.java:1375]: suspect you meant this?

```suggestion
            if (executor == EsExecutors.DIRECT_EXECUTOR_SERVICE && enableStackOverflowAvoidance) {
```

File: server/src/main/java/org/elasticsearch/transport/ForkingResponseHandlerRunnable.java [hunk 6]
```diff
@@ -30,7 +32,15 @@ abstract class ForkingResponseHandlerRunnable extends AbstractRunnable {
...
-        assert handler.executor() != EsExecutors.DIRECT_EXECUTOR_SERVICE : "forking handler required, but got " + handler;
+        this(handler, transportException, handler.executor());
+    }
+
+    ForkingResponseHandlerRunnable(
+        TransportResponseHandler<?> handler,
+        @Nullable TransportException transportException,
+        Executor executorUsed
+    ) {
+        assert executorUsed != EsExecutors.DIRECT_EXECUTOR_SERVICE : "forking handler required, but got " + handler;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/ForkingResponseHandlerRunnable.java:43]: I see, yeah this LGTM

File: server/src/main/java/org/elasticsearch/transport/TransportService.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/transport/TransportService.java:95]: LGTM

File: server/src/test/java/org/elasticsearch/transport/TransportServiceLifecycleTests.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/transport/TransportServiceLifecycleTests.java:255]: Ok I missed that, sorry for the noise.

File: server/src/main/java/org/elasticsearch/transport/TransportService.java (hunk 5 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/transport/TransportService.java]: yeah, sorry, should have run the tests :facepalm:

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by henningandersen** [comment]
Thanks David!

**Pull request merged**